llvm/lib/CodeGen/MachinePipeliner.cpp

```diff
@@ -2485,11 +2476,42 @@ bool SwingSchedulerDAG::schedulePipeline(SMSchedule &Schedule) {
     if (scheduleFound && LimitRegPressure)
       scheduleFound =
           !HRPDetector->detect(this, Schedule, Schedule.getMaxStageCount());
+
+    if (scheduleFound)
+      return true;
+  }
+
+  // Tried all node orders, no schedule was found.
+  return false;
+}
+
+/// Process the nodes in the computed order and create the pipelined schedule
+/// of the instructions, if possible. Return true if a schedule is found.
+bool SwingSchedulerDAG::schedulePipeline(SMSchedule &Schedule) {
+
+  if (NodeOrder.empty()) {
+    LLVM_DEBUG(dbgs() << "NodeOrder is empty! abort scheduling\n");
+    return false;
+  }
+  SmallVector<ArrayRef<SUnit *>, 4> Orders =
+      LoopPipelinerInfo->getNodeOrders(NodeOrder.getArrayRef(), Topo);
```

Comment:
After calculating the  SMS node order, the order is validated by `checkValidNodeOrder`. Do you think it tis necessary to validate the target node order as well?

Reply:
Good point. I had a look at what `checkValidNodeOrder` does, and it's really checking properties of the default swing modulo scheduling order that comes from the "circuit finding algorithm" described in some paper. (Don't ask me what that paper says, or what a circuit is please 🙏😄)

The important point is: the check can "fail" but this does not prevent the pipeliner from continuing, and this does not impact correctness.

Reply:
FYI: Getting a warning from that check correlates strongly with not finding a module schedule.  I think it indicates a property of the dependence graph which I will call 'openness': the graph without the recurrence edges can be drawn without intersections and without enclosing areas.